Arm VM Multicore Improvements

Apave24 opened the pull request:
These commits allow for more flexible core configurations for Arm VM systems

axel-h commented on libsel4vmmplatsupport/src/arch/arm/psci.c:
```diff
@@ -50,23 +50,37 @@ int handle_psci(vm_vcpu_t *vcpu, seL4_Word fn_number, bool convention)
                 smc_set_return_value(&regs, PSCI_INTERNAL_FAILURE);
             }
+        } else {
+            ZF_LOGE("Target CPU was an invalid number");
```
We should also log the invalid number
```suggestion
            ZF_LOGE("Target CPU is invalid: %d", target_vcpu->target_cpu);
```

Apave24 replied:
Done

axel-h commented on libsel4vmmplatsupport/src/arch/arm/psci.c:
```diff
@@ -50,23 +50,37 @@ int handle_psci(vm_vcpu_t *vcpu, seL4_Word fn_number, bool convention)
+            if ((selected_cpu >= 0) && start_new_vcpu(target_vcpu, entry_point_address, context_id, selected_cpu) == 0) {
                 smc_set_return_value(&regs, PSCI_SUCCESS);
             } else {
                 smc_set_return_value(&regs, PSCI_INTERNAL_FAILURE);
```
Let's add a message here also now that some auto-magic is happening here
```
ZF_LOGE("No unused physical core left for vCPU %u", vcpu->vcpu_id);
```

Apave24 replied:
Done

axel-h commented on libsel4vmmplatsupport/src/arch/arm/psci.c:
```diff
@@ -50,23 +50,37 @@ int handle_psci(vm_vcpu_t *vcpu, seL4_Word fn_number, bool convention)
             if (is_vcpu_online(target_vcpu)) {
                 smc_set_return_value(&regs, PSCI_ALREADY_ON);
+            } else if (start_new_vcpu(target_vcpu, entry_point_address, context_id, target_vcpu->target_cpu) == 0) {
+                smc_set_return_value(&regs, PSCI_SUCCESS);
```
Maybe log an error here also?
ZF_LOGE("[vCPU %u] could not start physical CPU %d , target_vcpu->target_cpu);

Apave24 replied:
Done in the else case

axel-h commented on libsel4vm/include/sel4vm/guest_vm_util.h:
```diff
@@ -72,6 +72,30 @@ static inline vm_vcpu_t *vm_find_free_unassigned_vcpu(vm_t *vm)
+ */
+static inline int vm_find_free_unassigned_cpu(vm_t *vm)
+{
+    for (int core = 0; core < vm->num_vcpus; core++) {
```
I'm not sure I get this algorithm, shouldn't we loop over the physical cores in the outer loop and check in the innter loop if this core is used by any VM?

Apave24 replied:
Agreed, CONFIG_MAX_NUM_NODES probably makes more sense here.

axel-h commented on libsel4vm/include/sel4vm/guest_vm_util.h:
```diff
@@ -72,6 +72,30 @@ static inline vm_vcpu_t *vm_find_free_unassigned_vcpu(vm_t *vm)
+                break;
+            }
+        }
+        if (core_allocated) {
```
reverse the ceck, then there is no need for the explicit `continue` and the algorithm is a bit easier to read
```suggestion
        if (!core_allocated) { 
            return core
        }
```

Apave24 replied:
Done

axel-h commented on libsel4vm/src/boot.c:
```diff
@@ -88,10 +88,17 @@ int vm_assign_vcpu_target(vm_vcpu_t *vcpu, int target_cpu)
     }
+
+#if CONFIG_MAX_NUM_NODES > 1
+    if (seL4_TCB_SetAffinity(vcpu->tcb.tcb.cptr, target_cpu)) {
```
Nitpicking: this just moves the check, but the logic is remains semantically broken here. It's easier to understand if this would be written as: 

```suggestion
    if (0 != seL4_TCB_SetAffinity(vcpu->tcb.tcb.cptr, target_cpu)) {
```
Or even more verbosely with an error mesage:
```
    int err = seL4_TCB_SetAffinity(vcpu->tcb.tcb.cptr, target_cpu);
    if (err) {
        ZF_LOGE("[vCPU %u] Failed to set CPU affinity to %d", target_cpu);
   }
```

Apave24 replied:
Done

axel-h commented on libsel4vmmplatsupport/src/arch/arm/psci.c:
```diff
@@ -50,23 +50,39 @@ int handle_psci(vm_vcpu_t *vcpu, seL4_Word fn_number, bool convention)
-            target_vcpu = vm_find_free_unassigned_vcpu(vcpu->vm);
-            if (target_vcpu && start_new_vcpu(target_vcpu, entry_point_address, context_id, target_cpu) == 0) {
+        vm_vcpu_t *target_vcpu = NULL;
+        if ((requested_cpu >= 0) && (requested_cpu < vcpu->vm->num_vcpus)) {
```
The `uintptr_t` can never be negative, so we just need to check the upper limit.

axel-h commented on libsel4vmmplatsupport/src/arch/arm/psci.c:
```diff
@@ -50,23 +50,39 @@ int handle_psci(vm_vcpu_t *vcpu, seL4_Word fn_number, bool convention)
+            if ((selected_cpu >= 0) && start_new_vcpu(target_vcpu, entry_point_address, context_id, selected_cpu) == 0) {
                 smc_set_return_value(&regs, PSCI_SUCCESS);
             } else {
+                ZF_LOGE("[vCPU %u] no unused physical core left", vcpu->vcpu_id);
```
...or starting failed. I think separating the errors is worth it, because that make debugging easier when something fails here or got out if sync somehow.

axel-h commented on libsel4vmmplatsupport/src/arch/arm/psci.c:
```diff
@@ -50,15 +50,24 @@ int handle_psci(vm_vcpu_t *vcpu, seL4_Word fn_number, bool convention)
         break;
     case PSCI_CPU_ON: {
-        uintptr_t target_cpu = smc_get_arg(&regs, 1);
+        uintptr_t requested_cpu = smc_get_arg(&regs, 1);
```
I think the guest would still pass MPIDR here, and that is not a linear number if there are multiple clusters to emulate. So we cannot use this as array index, but have to look it up in all vCPUs the guest got. We should really call even the variable `target_vmpidr` then.

axel-h commented on libsel4vm/src/boot.c:
```diff
@@ -92,6 +92,15 @@ int vm_assign_vcpu_target(vm_vcpu_t *vcpu, int target_cpu)
     }
+
+#if CONFIG_MAX_NUM_NODES > 1
+    int err = seL4_TCB_SetAffinity(vcpu->tcb.tcb.cptr, target_cpu);
```
I think to commit is ok. But it makes this generic code now, how do other architectures cope with this change then? Do we gave to change something in the x86 VMM part also?

Apave24 replied:
I believe `vm_assign_vcpu_target` is currently only used by ARM

axel-h commented:
Since this PR has multiple commits that seem independent, but have dependencies on other PRs, could you split this up that we can merge things independently and avoid unnecessary blocking?

Apave24 commented:
> Since this PR has multiple commits that seem independent, but have dependencies on other PRs, could you split this up that we can merge things independently and avoid unnecessary blocking?

Dropped commit `libsel4vm: guest cpu requests to vcpus from pcpus` from this PR and added to #85

Apave24 commented:
Bumping this PR. I believe everything has been addressed.

axel-h commented:
@Apave24 could you rebase this, then we also get another CI run there to see if this passes.

chrisguikema commented:
@axel-h @abrandnewusername @kent-mcleod 

Hey guys, just wanted to ping on this PR. This series of PRs makes it so that when the VM kicks off a secondary core, the VMM pins the VCPU to a PCPU defined in the `vmX.pcpus` list. I feel like this behavior makes the most sense given the current VMM configuration. Is this a good path forward? Can we work on getting this merged?
